core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,305 @@
+    "2. Archived to the target location." +
+    "3. Moved to the target location.";
+  protected static final String TARGET_FOLDER_DESCRIPTION = "Target folder path if user select action after process, " +
+    "either ARCHIVE or MOVE";
+  protected static final String FILES_TO_REPROCESS_DESCRIPTION = "Specifies whether the file(s) should be " +
+    "preprocessed.";
+  protected static final String TABLE_NAME_DESCRIPTION = "Name of the table to keep track of processed file(s).";
+
+  public static final Schema DEFAULT_XML_SCHEMA = Schema.recordOf(
+    "xmlSchema",
+    Schema.Field.of("offset", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("filename", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("record", Schema.of(Schema.Type.STRING))
+  );
+
+  private static final Gson GSON = new Gson();
+  private static final Type ARRAYLIST_PREPROCESSED_FILES  = new TypeToken<ArrayList<String>>() { }.getType();
+
+  private KeyValueTable processedFileTrackingTable;
+  private final XMLReaderConfig config;
+
+  public XMLReaderBatchSource(XMLReaderConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @VisibleForTesting
+  XMLReaderConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validateConfig();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_XML_SCHEMA);
+    if (StringUtils.isNotEmpty(config.tableName)) {
+      pipelineConfigurer.createDataset(config.tableName, KeyValueTable.class.getName());
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    if (StringUtils.isNotEmpty(config.pattern)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PATTERN, config.pattern);
+    }
+    XMLInputFormat.addInputPaths(job, config.path);
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_REPROCESSING_REQUIRED, this.config.reprocessingReq);
+    //set file tracking information in a temporary file, to be available to read outside plugin.
+    setFileTrackingInfo(context, conf);
+    if (StringUtils.isNotEmpty(this.config.actionAfterProcess)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION, this.config.actionAfterProcess);
+    }
+    if (StringUtils.isNotEmpty(this.config.actionAfterProcess)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER, this.config.targetFolder);
+    }
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH, this.config.nodePath);
+    XMLInputFormat.setInputPathFilter(job, BatchXMLFileFilter.class);
+    conf.set(XMLInputFormat.XML_INPUT_NAME_CONFIG, config.path);
+    context.setInput(Input.of(config.referenceName, new SourceInputFormatProvider(XMLInputFormat.class, conf)));
+  }
+
+  /**
+   * Method to set file tracking information.
+   * @param context
```

Comment:
lets remove these parameters and exception since you're not explaining anything about them. applies to the entire PR